### PR TITLE
It gets hot (Overheat mechanics for energy guns)

### PR DESCRIPTION
**Pull request by John-Candlebury**
#### Summary
Features  "Add overheat mechanics to energy guns."

#### Purpose of change

Energy weapons, with their lack of recoil and generally deep magazines are pretty hard to balance properly, especially as they become more powerfull, since the player can just stand somewhere and turn into a deadly accurate floodlamp of death after a few seconds of aiming.  Overheating is a good way to limit this ability.

Also anything close to a combat pulse laser would make tons of heat as I understand it.

#### Describe the solution

Guns can now define the fields `heat_per_shot` `overheat_threshold` and `cooling_value`  and the faults included here to display safe or potentially deadly overheat effects (which are mostly explained in the fault descriptions themselves). Overall, they can be used to create the standard overheat events of FPS's, cheap guns that melt during sustained fire,  and 40K style exploding plasma rifles.

Includes a demo Plasma gun for Aftershock and the vanilla lasers have been expanded to also overheat (mostly safely).

#### Testing
Used both overheat prone guns and normal ones.

### CODE REVIEW

**Comment by Raikiri** [comment]
Do vanilla laser rifles now overheat or does the change only affect them if aftershock is enabled? I surely hope it applies to vanilla.

**Comment by Profugo-Barbatus** [comment]
Love the sound of this. Now my energy weapons get downgraded to 'free, constrained problem solver' and I still break out the loaded mags and battle rattle if I'm planning to try and get into a mall or the like.

File: src/item.cpp [hunk 1]
```diff
@@ -6397,6 +6402,32 @@ std::string item::dirt_symbol() const
...
+    switch( std::min( 5, static_cast<int>( get_var( "gun_heat",
+                                           0 ) / ( type->gun->overheat_threshold ) * 5.0 ) ) ) {
+        case 1:
+            return "";
+        case 2:
+            return string_format( _( "<color_yellow>\u2583HEAT </color>" ) );
+        case 3:
+            return string_format( _( "<color_light_red>\u2585HEAT </color>" ) );
+        case 4:
+            return string_format( _( "<color_red>\u2587HEAT! </color>" ) );
```
Comment (ehughsbaird) [inline src/item.cpp:6423]: I believe this is how you can disable that.
```suggestion
            return string_format( _( "<color_red>\u2587HEAT! </color>" ) ); // NOLINT(cata-text-style)
```

**Comment by ilovehans10** [comment]
This makes me think of a mod for another game back in the day (might have been fallout), because they had a similar idea with a little twist.

They had a similar solution to this where laser weapons get hot, but they took it a step further and made it so that before the weapon would totally shut down it would first become inaccurate from heat. The reasoning was that the heat would warp focusing crystals and parts of the gun just enough to make it less accurate before it would just outright stop working. Idk how feasible that idea is, but I thought I should bring it up.

**Comment by SurFlurer** [comment]
Do you think RMES and CMES from mechs are immune from overheating? I guess they should also suffer from it.

**Comment by John-Candlebury** [comment]
Left them out since they have a whole mech attached that can potentially handle that better than just making the gun handheld. 

Afaik you cant fire them independently so I didnt consider adding overheats necessary.

**Comment by SurFlurer** [comment]
> Afaik you cant fire them independently so I didnt consider adding overheats necessary.

 They can be installed on a turret, making an extremely efficient killing machine.

**Pull request merged**